upb/decode_fast.c

```diff
@@ -0,0 +1,479 @@
+
+#undef F
+#undef FALLBACKS
+#undef FUNCNAME
+
+/* Generate all varint functions.
+ * {s,o,r} x {b1,v4,z4,v8,z8} x {1bt,2bt} */
+
+#define F(card, type, valbytes, tagbytes)                                      \
+  const char *upb_p##card##type##valbytes##_##tagbytes##bt(UPB_PARSE_PARAMS) { \
+    return fastdecode_varint(UPB_PARSE_ARGS, tagbytes, valbytes, CARD_##card,  \
+                             type##_ZZ, TABLENAME(type, valbytes));            \
+  }
+
+#define TYPES(card, tagbytes) \
+  F(card, b, 1, tagbytes)     \
+  F(card, v, 4, tagbytes)     \
+  F(card, v, 8, tagbytes)     \
+  F(card, z, 4, tagbytes)     \
+  F(card, z, 8, tagbytes)
+
+#define TAGBYTES(card) \
+  TYPES(card, 1)       \
+  TYPES(card, 2)
+
+TAGBYTES(s)
+TAGBYTES(o)
+/* TAGBYTES(r) */
+
+#undef z_ZZ
+#undef b_ZZ
+#undef v_ZZ
+#undef o_ONEOF
+#undef s_ONEOF
+#undef r_ONEOF
+#undef F
+#undef TYPES
+#undef TAGBYTES
+
+/* string fields **************************************************************/
+
+UPB_FORCEINLINE
+bool fastdecode_boundscheck(const char *ptr, size_t len, const char *end) {
+  uintptr_t uptr = (uintptr_t)ptr;
+  uintptr_t uend = (uintptr_t)end;
+  uintptr_t res = uptr + len;
+  return res < uptr || res > uend;
+}
+
+UPB_FORCEINLINE
+static const char *fastdecode_string(UPB_PARSE_PARAMS, int tagbytes,
+                                     upb_card card) {
+  upb_strview *dst;
+  const char *str;
+  int64_t len;
+  if (UPB_UNLIKELY(!fastdecode_checktag(data, tagbytes))) {
+    RETURN_GENERIC("string field tag mismatch\n");
+  }
+
+  dst = fastdecode_getfield(d, ptr, msg, &data, &hasbits, tagbytes,
+                            sizeof(upb_strview), card);
+  len = (int8_t)ptr[tagbytes];
+  str = ptr + tagbytes + 1;
+  dst->data = str;
+  dst->size = len;
+  if (UPB_UNLIKELY(fastdecode_boundscheck(str, len, d->limit))) {
+    dst->size = 0;
+    RETURN_GENERIC("string field len >1 byte\n");
+  }
+  return fastdecode_dispatch(d, str + len, msg, table, hasbits);
+}
+
+const char *upb_pss_1bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 1, CARD_s);
+}
+
+const char *upb_pos_1bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 1, CARD_o);
+}
+
+const char *upb_pss_2bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 2, CARD_s);
+}
+
+const char *upb_pos_2bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 2, CARD_o);
+}
+
+/* message fields *************************************************************/
+
+UPB_FORCEINLINE
+bool fastdecode_boundscheck2(const char *ptr, unsigned len, const char *end) {
+  uintptr_t uptr = (uintptr_t)ptr;
+  uintptr_t uend = (uintptr_t)end;
+  uintptr_t res = uptr + len;
+  return res < uptr || res > uend;
+}
+
+UPB_FORCEINLINE
+static const char *fastdecode_submsg(UPB_PARSE_PARAMS, int tagbytes,
+                                     int msg_ceil_bytes, upb_card card) {
+  const char *saved_limit = d->limit;
+  const char *saved_fastlimit = d->fastlimit;
```

Comment:
i don't know if saving fast limit is helpful balance between register pressure and not redoing computation

d->fastlimit = UPB_MIN(saved_limit, d->fastend); would recalculate it.

Reply:
But we compare against fastlimit in the loop for repeated fields:

```
    if (UPB_LIKELY(ptr < saved_fastlimit) &&
        fastdecode_readtag(ptr, tagbytes) == (uint16_t)data) {
```